conformance-test/retryStrategy.ts

```diff
@@ -0,0 +1,190 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {describe, it} from 'mocha';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as uuid from 'uuid';
+import pLimit = require('p-limit');
+import * as assert from 'assert';
+
+require('conformance-test/libraryMethods.ts');
+import {Bucket, File, Iam, Notification, Storage} from '../src/';
+
+interface RetryCase {
+  instructions: String[];
+}
+
+interface Method {
+  name: String;
+  resources: String[];
+}
+
+interface RetryTestCase {
+  id: number;
+  description: String;
+  retryCases: RetryCase[];
+  methods: Method[];
+  preconditionProvided: boolean;
+  expectSuccess: boolean;
+}
+
+interface MethodMap {
+  jsonApi: String;
+  nodejsStorageMethods: String[];
+}
+
+const testFile = fs.readFileSync(
+  //TODO change to require
+  path.join(
+    __dirname,
+    '../../conformance-test/test-data/retryStrategyTestData.json'
+  ),
+  'utf-8'
+);
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const testFileParsed = JSON.parse(testFile);
+const retryTestCases: RetryTestCase[] = testFileParsed.retryStrategyTests;
+
+const jsonToNodeApiMapping = fs.readFileSync(
+  //TODO change to require
+  path.join(
+    __dirname,
+    '../../conformance-test/test-data/retryInvocationMap.json'
+  ),
+  'utf-8'
+);
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const methodMap: Map<String, String[]> = JSON.parse(jsonToNodeApiMapping);
+const storage = new Storage(); //TODO: add apiEndpoint
+
+const TESTS_PREFIX = `storage-retry-tests-${shortUUID()}-`;
+const RETENTION_DURATION_SECONDS = 10;
+const OPTIONS = {
+  preconditionOpts: {
+    ifGenerationMatch: 100,
+    ifMetagenerationMatch: 100,
+  },
+};
+
+describe('retry conformance testing', () => {
+  for (
+    let testCaseIndex = 0;
+    testCaseIndex < retryTestCases.length;
+    testCaseIndex++
+  ) {
+    const testCase: RetryTestCase = retryTestCases[testCaseIndex];
+
+    describe(`Scenario ${testCase.id}`, () => {
+      excecuteScenario(testCase);
+    });
+  }
+});
+
+function excecuteScenario(testCase: RetryTestCase) {
+  testCase.retryCases.forEach((instructionSet: RetryCase) => {
+    configureTestBench(instructionSet.instructions);
+    testCase.methods.forEach(jsonMethod => {
+      const functionList = methodMap.get(jsonMethod.name);
+      functionList?.forEach(storageMethodString => {
+        const storageMethodObject = (global as any).storageMethodString;
```

Comment:
If `storageMethodString` is coming in as string type you may have to use bracket notation here instead.

Reply:
using bracket notation throws errors: "Type string cannot be used as an index type"
I will run the code and debug it from there.

Reply:
I would just make sure that using a `.` works with a string. Maybe TypeScript handles it cleaner than vanilla JS.

Reply:
Actually, I think it might be because `global` is of type `any` and not `object`. If you have a way to test, I would just make sure dot notation works as expected and leave it that way if so.

Reply:
I was thinking of testing it after we merge this PR? i'm sure i'll encounter other bugs along the way.